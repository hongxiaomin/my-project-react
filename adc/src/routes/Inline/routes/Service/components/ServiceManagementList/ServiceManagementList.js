import React from 'react'
import PropTypes from 'prop-types'
import WrapperFormattedMessage from 'components/WrapperFormattedMessage'
import Mqtt from 'mqtt'
import { connect } from 'react-redux'
import AlertMqttError from 'components/AlertMqttError'
import classes from './ServiceManagementList.scss'
import ServiceCard from '../ServiceCard'
import { updateServiceList, updateQueueStatus, addJobToQueueList } from '../../modules/Service'

export class ServiceManagementList extends React.Component {
  constructor(props) {
    super(props)
    this.client = null
    this.state = {
      isOpen: false,
      selectedIndex: -1,
      kernelIdArray: [],
      progress: '',
      mqttError: 'loading',
      visible: false,
    }
  }

  componentDidMount() {
    this.subscriptionMqttWithDeviceId()
  }

  componentWillUnmount() {
    // UnSubscribe mqtt
    if (this.client) {
      this.client.end(() => {
        console.log('Disconnect')
      })
    }
  }

  subscriptionMqttWithDeviceId = () => {
    let mqttClient
    if (__PROD__) {
      mqttClient = Mqtt.connect(`wss://${location.hostname}:59001`, {
        rejectUnauthorized: false,
        username: 'deltaww',
        password: 'Dare2Define',
      })
    } else {
      mqttClient = Mqtt.connect(__MQTT__, {
        rejectUnauthorized: false,
        username: 'deltaww',
        password: 'Dare2Define',
      })
    }

    mqttClient.stream.on('error', () => {
      this.setState({
        visible: true,
        mqttError: false,
      })
      mqttClient.end()
    })

    mqttClient.on('connect', () => {
      this.setState({
        mqttError: true,
      })
      mqttClient.subscribe('autojudge/service/queue')
      mqttClient.subscribe('autojudge/service/info')
    })

    mqttClient.on('message', (topic, message) => {
      let msgArray
      try {
        msgArray = JSON.parse(message.toString())
      } catch (e) {
        console.error('Json parsing fails！')
        return
      }

      // 新增一条队列信息
      if (topic === 'autojudge/service/queue') {
        if (msgArray.type === 'create') {
          this.props.addJobToQueueList(msgArray.data)
        } else if (msgArray.type === 'update') {
          this.props.updateQueueStatus(msgArray.data)
        }
        return
      }

      // 更新service卡片信息
      if (topic === 'autojudge/service/info') {
        this.props.updateServiceList(msgArray)
      }
    })

    this.client = mqttClient
  }

  handleMqttError = () => {
    this.setState({
      visible: false,
    })
  }

  render() {
    return (
      <div className={classes['ServiceManagementList-container']}>
        <section className={`${classes.title} ${classes.clear}`}>
          <h4 className={`${classes.left}`}><WrapperFormattedMessage id="nav.serviceManagement" /></h4>
        </section>
        <div className={classes.cards}>
          {
            this.props.serviceList.map(item => (<ServiceCard
              item={item}
              key={item.viewId}
              loading={this.props.loading.includes(item.serviceId)}
            />))
          }
        </div>
        <AlertMqttError
          isOpen={this.state.visible}
          isError={this.state.mqttError}
          handleCancel={this.handleMqttError}
        />
      </div>
    )
  }
}

ServiceManagementList.propTypes = {
  updateServiceList: PropTypes.func.isRequired,
  updateQueueStatus: PropTypes.func.isRequired,
  addJobToQueueList: PropTypes.func.isRequired,
  serviceList: PropTypes.array.isRequired,
  loading: PropTypes.array.isRequired,
}

export default connect(
  ({ auth, service, intl }) => ({
    intl,
    token: auth.token,
    serviceList: service.serviceList,
    loading: service.loading.service,
  }), {
    updateServiceList,
    updateQueueStatus,
    addJobToQueueList,
  })(ServiceManagementList)
