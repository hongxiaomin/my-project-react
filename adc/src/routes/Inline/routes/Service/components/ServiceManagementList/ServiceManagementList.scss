.ServiceManagementList-container {
  width: 45%;

  // float: left;
  // overflow: hidden;
  // border: 1px solid #0087DC;
  border-radius: 5px;
  padding: 0 0 10px 0;
  // height: 415px;
}
.cards {
  display: flex;
  flex-basis: auto;
  justify-content: space-center;
  flex-wrap: wrap;
}
  // .clear{
  //
  //   &:after{
  //     content: '';
  //     display: block;
  //     clear: both;
  //   }
  // }
  // .right{
  //   float: right;
  // }
  // .left{
  //   float: left;
  // }
  // .mr10{
  //   margin-right: 10px;
  // }
  // .ml10{
  //   margin-left: 10px;
  // }
  // .c1{
  //   color: #0087DC;
  // }
  // .cn{
  //   text-align: center;
  // }
  //
  // .action{
  //   position: absolute;
  //   top: 30px;
  //   right: 140px;
  // }
  //
  // .eye{
  //   position: absolute;
  //   top: 30px;
  //   right: 82px;
  // }
  //
  // .log{
  //   position: absolute;
  //   top: 30px;
  //   right: 20px;
  // }

  .title{
    // background: #F7F7F7;
    padding: 10px;
    position: relative;
  }

//   .ag-table{
//     height: 330px;
//   }
//
//   .cell {
//     border-radius: 5px;
//     color: #fff;
//     background: #9B9B9B;
//     display: inline-block;
//     width: 100px;
//     overflow: hidden;
//     font-size: 12px
//   }
//
//   .progress {
//     width: initial;
//     margin-left: 5px;
//     padding: 0 5px;
//     background: #2887C3;
//     border-radius: 0;
//   }
//
//   .status{
//     padding: 5px 10px;
//     border-radius: 5px 0 5px 0;
//     color: #fff;
//     background: #9B9B9B;
//     &.status-s{
//       background: #7ED321;
//     }
//
//     &.status-d{
//       background: #D0021B;
//     }
//
//     &.status-w{
//       background: #F5A623;
//     }
//
//     &.status-p{
//       background: #4A90E2;
//     }
//
//   }
// }
