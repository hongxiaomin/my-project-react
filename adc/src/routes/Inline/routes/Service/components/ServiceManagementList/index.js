import ServiceManagementListContainer from './ServiceManagementList'

export default ServiceManagementListContainer
