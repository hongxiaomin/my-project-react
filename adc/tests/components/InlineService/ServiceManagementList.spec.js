import React from 'react'
import { ServiceManagementList } from 'routes/Inline/routes/Service/components/ServiceManagementList/ServiceManagementList'

describe('(Component) ServiceManagementList', () => {
  const updateServiceList = jest.fn()
  const updateQueueStatus = jest.fn()
  const addJobToQueueList = jest.fn()
  const serviceList = [{ id: 1 }, { id: 2 }]
  const loading = []
  it('render normally', () => {
    const wrapper = shallow(
      <ServiceManagementList
        updateServiceList={updateServiceList}
        updateQueueStatus={updateQueueStatus}
        addJobToQueueList={addJobToQueueList}
        serviceList={serviceList}
        loading={loading}
      />,
    )
    // wrapper.find('Button').last().simulate('click')
    expect(wrapper).toMatchSnapshot()
  })
})