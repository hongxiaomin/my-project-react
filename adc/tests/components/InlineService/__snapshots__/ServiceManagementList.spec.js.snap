// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) ServiceManagementList render normally 1`] = `
<div
  className="ServiceManagementList-container"
>
  <section
    className="title clear"
  >
    <h4
      className="left"
    >
      Service Management
    </h4>
  </section>
  <div
    className="cards"
  >
    <Connect(ServiceCard)
      item={
        Object {
          "id": 1,
        }
      }
      loading={false}
    />
    <Connect(ServiceCard)
      item={
        Object {
          "id": 2,
        }
      }
      loading={false}
    />
  </div>
  <AlertMqttError
    handleCancel={[Function]}
    isError="loading"
    isOpen={false}
  />
</div>
`;
